File: packages/agent-mesh/sdks/go/identity.go
package agentmesh

import (
	"crypto/ed25519"
	"crypto/rand"
	"encoding/json"
	"fmt"
)

// AgentIdentity holds an agent's DID and Ed25519 key pair.
type AgentIdentity struct {
	DID          string              `json:"did"`
	PublicKey    ed25519.PublicKey    `json:"public_key"`
	Capabilities []string            `json:"capabilities,omitempty"`
	privateKey   ed25519.PrivateKey
}

// GenerateIdentity creates a new Ed25519-based identity for the given agent.
func GenerateIdentity(agentID string, capabilities []string) (*AgentIdentity, error) {
	pub, priv, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		return nil, fmt.Errorf("generating key pair: %w", err)
	}
	return &AgentIdentity{
		DID:          fmt.Sprintf("did:agentmesh:%s", agentID),
		PublicKey:    pub,
		Capabilities: capabilities,
		privateKey:   priv,
	}, nil
}

// Sign signs data with the agent's private key.
func (a *AgentIdentity) Sign(data []byte) ([]byte, error) {
	if a.privateKey == nil {
		return nil, fmt.Errorf("no private key available")
	}
	return ed25519.Sign(a.privateKey, data), nil
}

// Verify checks a signature against data using the agent's public key.
func (a *AgentIdentity) Verify(data, signature []byte) bool {
	return ed25519.Verify(a.PublicKey, data, signature)
}

// identityJSON is used for JSON marshalling (excludes private key).
type identityJSON struct {
	DID          string   `json:"did"`
	PublicKey    []byte   `json:"public_key"`
	Capabilities []string `json:"capabilities,omitempty"`
}

// ToJSON serialises the public portion of the identity.
func (a *AgentIdentity) ToJSON() ([]byte, error) {
	return json.Marshal(identityJSON{
		DID:          a.DID,
		PublicKey:    []byte(a.PublicKey),
		Capabilities: a.Capabilities,
	})
}

// FromJSON deserialises an identity from JSON (public fields only).
func FromJSON(data []byte) (*AgentIdentity, error) {
	var j identityJSON
	if err := json.Unmarshal(data, &j); err != nil {
		return nil, fmt.Errorf("unmarshalling identity: %w", err)
	}
	return &AgentIdentity{
		DID:          j.DID,
		PublicKey:    ed25519.PublicKey(j.PublicKey),
		Capabilities: j.Capabilities,
	}, nil
}

File: packages/agent-mesh/sdks/typescript/src/identity.ts
import { AgentIdentityJSON } from './types';
import { createHash, randomBytes, sign, verify, generateKeyPairSync } from 'crypto';

/**
 * Agent identity built on Ed25519 key pairs and DID identifiers.
 */
export class AgentIdentity {
  readonly did: string;
  readonly publicKey: Uint8Array;
  private readonly _privateKey: Uint8Array;
  private readonly _capabilities: string[];

  private constructor(
    did: string,
    publicKey: Uint8Array,
    privateKey: Uint8Array,
    capabilities: string[],
  ) {
    this.did = did;
    this.publicKey = publicKey;
    this._privateKey = privateKey;
    this._capabilities = capabilities;
  }

  /** Generate a new agent identity with an Ed25519 key pair. */
  static generate(agentId: string, capabilities: string[] = []): AgentIdentity {
    const { publicKey, privateKey } = generateKeyPairSync('ed25519');

    const pubBytes = new Uint8Array(
      publicKey.export({ type: 'spki', format: 'der' }),
    );
    const privBytes = new Uint8Array(
      privateKey.export({ type: 'pkcs8', format: 'der' }),
    );

    const fingerprint = createHash('sha256')
      .update(pubBytes)
      .digest('hex')
      .slice(0, 16);

    const did = `did:agentmesh:${agentId}:${fingerprint}`;

    return new AgentIdentity(did, pubBytes, privBytes, capabilities);
  }

  /** Sign arbitrary data and return the signature bytes. */
  sign(data: Uint8Array): Uint8Array {
    const privateKeyObject = require('crypto').createPrivateKey({
      key: Buffer.from(this._privateKey),
      format: 'der',
      type: 'pkcs8',
    });
    const sig = sign(null, Buffer.from(data), privateKeyObject);
    return new Uint8Array(sig);
  }

  /** Verify a signature against this identity's public key. */
  verify(data: Uint8Array, signature: Uint8Array): boolean {
    try {
      const publicKeyObject = require('crypto').createPublicKey({
        key: Buffer.from(this.publicKey),
        format: 'der',
        type: 'spki',
      });
      return verify(null, Buffer.from(data), publicKeyObject, Buffer.from(signature));
    } catch {
      return false;
    }
  }

  /** Serialize to a plain JSON-safe object. */
  toJSON(): AgentIdentityJSON {
    return {
      did: this.did,
      publicKey: Buffer.from(this.publicKey).toString('base64'),
      privateKey: Buffer.from(this._privateKey).toString('base64'),
      capabilities: [...this._capabilities],
    };
  }

  /** Reconstruct an AgentIdentity from its JSON representation. */
  static fromJSON(json: AgentIdentityJSON): AgentIdentity {
    const pubKey = new Uint8Array(Buffer.from(json.publicKey, 'base64'));
    const privKey = json.privateKey
      ? new Uint8Array(Buffer.from(json.privateKey, 'base64'))
      : new Uint8Array(0);
    return new AgentIdentity(json.did, pubKey, privKey, json.capabilities ?? []);
  }

  get capabilities(): readonly string[] {
    return this._capabilities;
  }
}
